File: backend/src/routes/clients.js
const express = require('express');
const { getDatabase } = require('../database/init');
const { authenticateUser } = require('../middleware/auth');
const { clientSchema, updateClientSchema } = require('../validation/schemas');
const { parseId, sendDbError, buildUpdateClause } = require('../utils/routeHelpers');
const { CLIENT_COLUMNS } = require('../database/queries');

const router = express.Router();

// All routes require authentication
router.use(authenticateUser);

// Get all clients for authenticated user
router.get('/', (req, res) => {
  const db = getDatabase();
  
  db.all(
    `SELECT ${CLIENT_COLUMNS} FROM clients WHERE user_email = ? ORDER BY name`,
    [req.userEmail],
    (err, rows) => {
      if (err) {
        return sendDbError(res, err);
      }
      
      res.json({ clients: rows });
    }
  );
});

// Get specific client
router.get('/:id', (req, res) => {
  const clientId = parseId(req.params.id);
  
  if (clientId === null) {
    return res.status(400).json({ error: 'Invalid client ID' });
  }
  
  const db = getDatabase();
  
  db.get(
    `SELECT ${CLIENT_COLUMNS} FROM clients WHERE id = ? AND user_email = ?`,
    [clientId, req.userEmail],
    (err, row) => {
      if (err) {
        return sendDbError(res, err);
      }
      
      if (!row) {
        return res.status(404).json({ error: 'Client not found' });
      }
      
      res.json({ client: row });
    }
  );
});

// Create new client
router.post('/', (req, res, next) => {
  try {
    const { error, value } = clientSchema.validate(req.body);
    if (error) {
      return next(error);
    }

    const { name, description, department, email } = value;
    const db = getDatabase();

    db.run(
      'INSERT INTO clients (name, description, department, email, user_email) VALUES (?, ?, ?, ?, ?)',
      [name, description || null, department || null, email || null, req.userEmail],
      function(err) {
        if (err) {
          return sendDbError(res, err, 'Failed to create client');
        }

        // Return the created client
        db.get(
          `SELECT ${CLIENT_COLUMNS} FROM clients WHERE id = ?`,
          [this.lastID],
          (err, row) => {
            if (err) {
              return sendDbError(res, err, 'Client created but failed to retrieve');
            }

            res.status(201).json({ 
              message: 'Client created successfully',
              client: row 
            });
          }
        );
      }
    );
  } catch (error) {
    next(error);
  }
});

// Update client
router.put('/:id', (req, res, next) => {
  try {
    const clientId = parseId(req.params.id);
    
    if (clientId === null) {
      return res.status(400).json({ error: 'Invalid client ID' });
    }

    const { error, value } = updateClientSchema.validate(req.body);
    if (error) {
      return next(error);
    }

    const db = getDatabase();

    // Check if client exists and belongs to user
    db.get(
      'SELECT id FROM clients WHERE id = ? AND user_email = ?',
      [clientId, req.userEmail],
      (err, row) => {
        if (err) {
          return sendDbError(res, err);
        }

        if (!row) {
          return res.status(404).json({ error: 'Client not found' });
        }

        // Build update query dynamically
        const { setClause, params } = buildUpdateClause([
          { column: 'name', value: value.name, include: value.name !== undefined },
          { column: 'description', value: value.description || null, include: value.description !== undefined },
          { column: 'department', value: value.department || null, include: value.department !== undefined },
          { column: 'email', value: value.email || null, include: value.email !== undefined }
        ]);
        params.push(clientId, req.userEmail);

        const query = `UPDATE clients SET ${setClause} WHERE id = ? AND user_email = ?`;

        db.run(query, params, function(err) {
          if (err) {
            return sendDbError(res, err, 'Failed to update client');
          }

          // Return updated client
          db.get(
            `SELECT ${CLIENT_COLUMNS} FROM clients WHERE id = ?`,
            [clientId],
            (err, row) => {
              if (err) {
                return sendDbError(res, err, 'Client updated but failed to retrieve');
              }

              res.json({
                message: 'Client updated successfully',
                client: row
              });
            }
          );
        });
      }
    );
  } catch (error) {
    next(error);
  }
});

// Delete all clients for authenticated user
router.delete('/', (req, res) => {
  const db = getDatabase();
  
  db.run(
    'DELETE FROM clients WHERE user_email = ?',
    [req.userEmail],
    function(err) {
      if (err) {
        return sendDbError(res, err, 'Failed to delete clients');
      }
      
      res.json({ 
        message: 'All clients deleted successfully',
        deletedCount: this.changes
      });
    }
  );
});

// Delete client
router.delete('/:id', (req, res) => {
  const clientId = parseId(req.params.id);
  
  if (clientId === null) {
    return res.status(400).json({ error: 'Invalid client ID' });
  }
  
  const db = getDatabase();
  
  // Check if client exists and belongs to user
  db.get(
    'SELECT id FROM clients WHERE id = ? AND user_email = ?',
    [clientId, req.userEmail],
    (err, row) => {
      if (err) {
        return sendDbError(res, err);
      }
      
      if (!row) {
        return res.status(404).json({ error: 'Client not found' });
      }
      
      // Delete client (work entries will be deleted due to CASCADE)
      db.run(
        'DELETE FROM clients WHERE id = ? AND user_email = ?',
        [clientId, req.userEmail],
        function(err) {
          if (err) {
            return sendDbError(res, err, 'Failed to delete client');
          }
          
          res.json({ message: 'Client deleted successfully' });
        }
      );
    }
  );
});

module.exports = router;

File: backend/src/routes/clients.py
from fastapi import APIRouter, Depends, HTTPException

from src.database import get_connection
from src.dependencies import get_authenticated_user
from src.schemas import CreateClientRequest, UpdateClientRequest

router = APIRouter(prefix="/api/clients", tags=["clients"])


@router.get("/")
def get_clients(user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id, name, description, department, email, created_at, updated_at "
                "FROM clients WHERE user_email = %s ORDER BY name",
                (user_email,),
            )
            return {"clients": cur.fetchall()}
    finally:
        conn.close()


@router.get("/{client_id}")
def get_client(client_id: int, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id, name, description, department, email, created_at, updated_at "
                "FROM clients WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            row = cur.fetchone()
            if not row:
                raise HTTPException(status_code=404, detail="Client not found")
            return {"client": row}
    finally:
        conn.close()


@router.post("/", status_code=201)
def create_client(body: CreateClientRequest, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "INSERT INTO clients (name, description, department, email, user_email) "
                "VALUES (%s, %s, %s, %s, %s)",
                (body.name, body.description, body.department, body.email, user_email),
            )
            new_id = cur.lastrowid
            cur.execute(
                "SELECT id, name, description, department, email, created_at, updated_at "
                "FROM clients WHERE id = %s",
                (new_id,),
            )
            return {"message": "Client created successfully", "client": cur.fetchone()}
    finally:
        conn.close()


@router.put("/{client_id}")
def update_client(
    client_id: int,
    body: UpdateClientRequest,
    user_email: str = Depends(get_authenticated_user),
):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM clients WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail="Client not found")

            data = body.model_dump(exclude_unset=True)
            if not data:
                raise HTTPException(status_code=400, detail="No fields to update")

            cur.execute(
                "SELECT id, name, description, department, email FROM clients "
                "WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            current = cur.fetchone()

            cur.execute(
                "UPDATE clients SET name = %s, description = %s, department = %s, "
                "email = %s, updated_at = CURRENT_TIMESTAMP "
                "WHERE id = %s AND user_email = %s",
                (
                    data.get("name", current["name"]),
                    data["description"] if "description" in data else current["description"],
                    data["department"] if "department" in data else current["department"],
                    data["email"] if "email" in data else current["email"],
                    client_id,
                    user_email,
                ),
            )
            cur.execute(
                "SELECT id, name, description, department, email, created_at, updated_at "
                "FROM clients WHERE id = %s",
                (client_id,),
            )
            return {"message": "Client updated successfully", "client": cur.fetchone()}
    finally:
        conn.close()


@router.delete("/")
def delete_all_clients(user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute("DELETE FROM clients WHERE user_email = %s", (user_email,))
            return {"message": "All clients deleted successfully", "deletedCount": cur.rowcount}
    finally:
        conn.close()


@router.delete("/{client_id}")
def delete_client(client_id: int, user_email: str = Depends(get_authenticated_user)):
    conn = get_connection()
    try:
        with conn.cursor() as cur:
            cur.execute(
                "SELECT id FROM clients WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            if not cur.fetchone():
                raise HTTPException(status_code=404, detail="Client not found")

            cur.execute(
                "DELETE FROM clients WHERE id = %s AND user_email = %s",
                (client_id, user_email),
            )
            return {"message": "Client deleted successfully"}
    finally:
        conn.close()
